### ISSUE TITLE
Multistage first WORKDIR is present in scratch container

### ISSUE THREAD

**Issue by decoomanj** [moby/moby#36607]
**Description**

When creating a scratch container in a multistage build, the WORKDIR from the first container is present in the scratch container. This does not happen with non-scratch containers. 

**Steps to reproduce the issue:**
1. Create the following Dockerfile

```docker
FROM golang:1-alpine as test
RUN mkdir /test
WORKDIR /test
RUN echo "package main" > test.go
RUN echo "func main() { for true{} }" >> test.go
RUN ls && go build -ldflags "-s" test.go

FROM scratch
COPY --from=test /test/test /app
ENTRYPOINT ["/app"]
```

2. Build the dockerfile: `docker build -t jan .`

3. Run the container: 
`docker run --rm  -d --name jan1 jan:latest*`

4. Attach a debian to inspect the filesystem of *jan1*:

`docker run --rm -ti  --pid container:jan1 --name jan2 --cap-add sys_ptrace  debian`

5. Go to `/proc/1/root`. You will see the directory `/test` there:

```shell
-rwxr-xr-x   1 root root 689600 Mar 15 10:27 app
drwxr-xr-x   5 root root    340 Mar 15 10:35 dev
drwxr-xr-x   2 root root   4096 Mar 15 10:35 etc
dr-xr-xr-x 360 root root      0 Mar 15 10:35 proc
dr-xr-xr-x  13 root root      0 Mar 15 10:35 sys
drwxr-xr-x   2 root root   4096 Mar 15 10:27 test
```

I assume the test-directory should not be there. When using `debian` the `test`  directory is not there:

1. Change the Dockerfile to:
```docker
FROM golang:1-alpine as test
RUN mkdir /test
WORKDIR /test
RUN echo "package main" > test.go
RUN echo "func main() { for true{} }" >> test.go
RUN ls && go build -ldflags "-s" test.go

FROM debian
COPY --from=test /test/test /app
ENTRYPOINT ["/app"]
```
2. Look at the root filesystem in the Debian container. There is NO `/test` directory.

**Describe the results you received:**

Different behaviour between "scratch" and for example "debian"

**Describe the results you expected:**

The behaviour should be identical. The WORKDIR is there or it is not there.

**Additional information you deem important (e.g. issue happens only occasionally):**

You can fix the scratch container by adding `WORKDIR /` to the scratch image. But this seems not very intuitive. Maybe the behaviour is expected, but it is then not good documented.

**Output of `docker version`:**

```
Client:
 Version:      17.05.0-ce
 API version:  1.29
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:10:54 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:10:54 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**

```
Containers: 11
 Running: 1
 Paused: 0
 Stopped: 10
Images: 1426
Server Version: 17.05.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 976
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9048e5e50717ea4497b757314bad98ea3763c145
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.13.0-36-generic
Operating System: Ubuntu 17.10
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 30.85GiB
Name: mordor
ID: YTRI:MVWK:R75A:22SH:XLST:ECLG:46OQ:ORW3:Z32Q:RLW4:VPQS:P3J4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**REMARKS:**

When the first multistage container sets a WORKDIR, this WORKDIR is used in the following scratch container. You can check this with golang `os.Executable()` and `os.Getwd()`. The `Getwd()` returns `/test` in the scratch container, which was the WORKDIR of the builder! Like this, relative filereferences do not work. A workaround here again is setting WORKDIR to / in the scratch container explicitly.

**WORKAROUND**

With this workaround, the `/test` directory is no longer in the scratch container:

```docker
FROM golang:1-alpine as test
RUN mkdir /test
WORKDIR /test
RUN echo "package main" > test.go
RUN echo "func main() { for true{} }" >> test.go
RUN ls && go build -ldflags "-s" test.go

FROM scratch
WORKDIR /
COPY --from=test /test/test /app
ENTRYPOINT ["/app"]
```

**Comment by vdemeester** [comment]
cc  @dnephin @tonistiigi

**Comment by tonistiigi** [comment]
I think this was fixed around `v17.07`. Please check a newer version.

**Comment by decoomanj** [comment]
@tonistiigi Tried to reproduce the error with 17.12.1-ce. Looks fine there.

**Comment by thaJeztah** [comment]
Yes, this looks to be fixed in Docker 17.06 (can find the PR)


Simplified reproduction steps:

```bash
docker build -t repro-36607 --no-cache -<<'EOF'
FROM busybox AS test
RUN mkdir /test
WORKDIR /test

FROM scratch
COPY --from=test /bin/busybox /busybox
ENTRYPOINT ["/busybox"]
EOF
```


on 17.05;


```bash
docker run --rm repro-36607 ls -la /

total 1060
drwxr-xr-x    1 0        0             4096 Mar 16 09:14 .
drwxr-xr-x    1 0        0             4096 Mar 16 09:14 ..
-rwxr-xr-x    1 0        0                0 Mar 16 09:14 .dockerenv
-rwxr-xr-x    1 0        0          1067344 Feb 20 17:57 busybox
drwxr-xr-x    5 0        0              340 Mar 16 09:14 dev
drwxr-xr-x    2 0        0             4096 Mar 16 09:14 etc
dr-xr-xr-x  197 0        0                0 Mar 16 09:14 proc
dr-xr-xr-x   13 0        0                0 Mar 16 09:14 sys
drwxr-xr-x    2 0        0             4096 Mar 16 09:14 test
```

on 17.06:

```bash
/ # docker run --rm repro-36607 ls -la /
total 1056
drwxr-xr-x    1 0        0             4096 Mar 16 09:16 .
drwxr-xr-x    1 0        0             4096 Mar 16 09:16 ..
-rwxr-xr-x    1 0        0                0 Mar 16 09:16 .dockerenv
-rwxr-xr-x    1 0        0          1067344 Feb 20 17:57 busybox
drwxr-xr-x    5 0        0              340 Mar 16 09:16 dev
drwxr-xr-x    2 0        0             4096 Mar 16 09:16 etc
dr-xr-xr-x  198 0        0                0 Mar 16 09:16 proc
dr-xr-xr-x   13 0        0                0 Mar 16 09:16 sys
```

**Comment by thaJeztah** [comment]
I _think_ this was the one fixing it; https://github.com/moby/moby/pull/33179 (see related issues https://github.com/moby/moby/issues/33176, https://github.com/linuxkit/linuxkit/issues/1807, https://github.com/moby/moby/issues/33550)

Closing this one because it's fixed, but feel free to continue the conversation
